module/zfs/zio.c

```diff
@@ -2015,8 +2015,12 @@ zio_taskq_member(zio_t *zio, zio_taskq_type_t q)
 static zio_t *
 zio_issue_async(zio_t *zio)
 {
+	/* ref: ZIO_STAGE_ISSUE_ASYNC for other types */
+	if ((zio->io_type == ZIO_TYPE_WRITE) &&
+	    !(zio->io_flags & ZIO_FLAG_GANG_CHILD)) {
+		spa_select_allocator(zio);
+	}
```

Comment:
Formally allocation and issue_async are different pipeline stages.  In theory there may be ZIO types that have allocation, but do not have issue_async.  So this may be not the best place to do it.  Check for GANG_CHILD also looks like a workaround.

Reply:
I have re-worked to use zio_wait/nowait for allocator selection, and to make gang child zios inherit their allocator from their pio.